restli-tools/src/main/java/com/linkedin/restli/tools/snapshot/gen/PegasusSchemaSnapshotExporter.java

```diff
@@ -0,0 +1,159 @@
+import com.linkedin.data.schema.DataSchema;
+import com.linkedin.data.schema.DataSchemaResolver;
+import com.linkedin.data.schema.NamedDataSchema;
+import com.linkedin.data.schema.SchemaToPdlEncoder;
+import com.linkedin.data.schema.grammar.PdlSchemaParser;
+import com.linkedin.data.schema.resolver.MultiFormatDataSchemaResolver;
+import com.linkedin.pegasus.generator.GeneratorResult;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.commons.io.FileUtils;
+
+/**
+ * PegasusSchemaSnapshotExporter, generating pegasus schema snapshot(.pdl) files
+ *
+ * @author Yingjie Bi
+ */
+public class PegasusSchemaSnapshotExporter
+{
+  private static final String PEGASUS_SCHEMA_SNAPSHOT_SUFFIX = ".pdl";
+
+  private static final String PDL = "pdl";
+
+  /**
+   * Generate pegasus schema snapshot(pegasusSchemaSnapshot.pdl) files to the provided output directory
+   * based on the given input pegasus schemas.
+   *
+   * @param resolverPath schema resolver path
+   * @param inputPath input files directory
+   * @param outputDir output files directory
+   * @return GeneratorResult
+   * @throws IOException
+   */
+  public GeneratorResult export(String resolverPath, String inputPath, File outputDir) throws IOException {
+    List<DataSchema> dataSchemas = parseDataSchema(resolverPath, inputPath);
+    pegasusSchemaSnapshotResult result = new pegasusSchemaSnapshotResult();
+    for (DataSchema dataSchema : dataSchemas)
+    {
+     File file = writeSnapshotFile(outputDir, ((NamedDataSchema) dataSchema).getFullName(), dataSchema);
+     result.addModifiedFile(file);
+     result.addTargetFile(file);
+    }
+    return result;
+  }
+
+  private static List<DataSchema> parseDataSchema(String resolverPath, String inputPath)
+      throws RuntimeException, FileNotFoundException
+  {
+    Iterator<File> iterator = FileUtils.iterateFiles(new File(inputPath), new String[]{PDL}, true);
+    DataSchemaResolver resolver = MultiFormatDataSchemaResolver.withBuiltinFormats(resolverPath);
+    List<DataSchema> schemas = new ArrayList<>();
+    while (iterator.hasNext())
+    {
+      File inputFile = iterator.next();
+      PdlSchemaParser parser = new PdlSchemaParser(resolver);
+      parser.parse(new FileInputStream(inputFile));
+      if (parser.hasError())
+      {
+        throw new RuntimeException(parser.errorMessage());
+      }
+
+      List<DataSchema> topLevelDataSchemas = parser.topLevelDataSchemas();
+      if (topLevelDataSchemas == null || topLevelDataSchemas.isEmpty() || topLevelDataSchemas.size() > 1)
+      {
+        throw new RuntimeException("Could not parse extension schema : " + inputFile.getAbsolutePath());
+      }
+      DataSchema topLevelDataSchema = topLevelDataSchemas.get(0);
```

Comment:
Is `get(0)` meant to get top level schema only?
If say sub-level schema changed, and we just `get(0)`, that means we dont treat sub-level schema change as change? Since the snapshot didn't change.
By sub-level I mean some recursive entity in the top level record's field

Reply:
thanks for addressing this question offline